drivers/hmis/app/graphql/mutations/bulk_review_external_submissions.rb

```diff
@@ -22,36 +22,39 @@ def resolve(external_submission_ids:)
       access_denied! unless current_permission?(permission: :can_manage_external_form_submissions, entity: project)
       access_denied! if definitions.any?(&:updates_client_or_enrollment?) && !current_permission?(permission: :can_edit_enrollments, entity: project)
 
-      already_reviewed = submissions.filter { |s| s.enrollment_id.present? || s.status == 'reviewed' }
-      error_out('Submissions are already processed: ' + already_reviewed.pluck(:id).inspect) if already_reviewed.any?
-
       should_auto_enter = project.should_auto_enter?
 
-      HmisExternalApis::ExternalForms::FormSubmission.transaction do
-        submissions.each do |record|
-          record.status = 'reviewed'
-          record.run_form_processor(current_user, project: project)
+      # maps record IDs to error messages
+      failed_to_review = {}
+
+      submissions.each do |record|
+        raise ArgumentError, 'Already reviewed' if record.enrollment_id.present? || record.status == 'reviewed'
 
-          if record.enrollment&.new_record?
-            error_out(record.enrollment.client.errors.full_messages) unless record.enrollment.client.valid?
-            error_out(record.enrollment.errors.full_messages) unless record.enrollment.valid?
+        record.status = 'reviewed'
+        record.run_form_processor(current_user, project: project)
 
-            record.enrollment.client.save!
-            should_auto_enter ? record.enrollment.save_and_auto_enter! : record.enrollment.save_in_progress!
-          end
+        if record.enrollment&.new_record?
+          raise ArgumentError, record.enrollment.client.errors.full_messages unless record.enrollment.client.valid?
+          raise ArgumentError, record.enrollment.errors.full_messages unless record.enrollment.valid?
 
-          record.form_processor.save!
-          record.save!
+          record.enrollment.client.save!
+          should_auto_enter ? record.enrollment.save_and_auto_enter! : record.enrollment.save_in_progress!
         end
-      end
 
-      { success: true }
-    end
+        record.form_processor.save!
+        record.save!
+      rescue StandardError => e
```

Comment:
Rescuing `StandardError` can cause unexpected behavior, because I most ruby errors are subclasses of StandardError. So basically _any_ execution error that happens in the block above will be swallowed.

A safer option is to just skip and count up failures as you go:

```ruby
submissions.each do |record|
    next if already_processed

    if !record.valid?
       failed_to_review << record
       next
    end

end
```

Reply:
Ok, sounds good! Confirming behavior change - it's fine if a submission has already been reviewed (we should just silently pass on rather than returning it as a failure?) I like this change, just want to make sure it's also what you intended.

Reply:
Yeah I think that's good, it mirrors how bulk services works now, too. It will help in situations where there are multiple users reviewing at the same time.

Reply:
Hm so, I think we still need some sort of try/catch here because the failure won't always be due to validation errors on the record or its related record. The reason I originally caught `StandardError` is bc that's what's [caught and then re-thrown by the FormProcessor](https://github.com/greenriver/hmis-warehouse/blob/5eed1631d1ae87951eb01bb1b4d4e7d064142b27/drivers/hmis/app/models/hmis/form/form_processor.rb#L80-L84). We could catch `RuntimeError` specifically, or modify [unknown_field_error](https://github.com/greenriver/hmis-warehouse/blob/5eed1631d1ae87951eb01bb1b4d4e7d064142b27/drivers/hmis/app/models/hmis/form/form_processor.rb#L51-L53) to return a different, more specific type of error.

Reply:
I guess my proposal is to catch `RuntimeError` and, since we need the try/catch anyway, to reduce code cruft by raising if the client or enrollment were invalid also (instead of your suggestion of `if valid?`, but I could go either way on this).  https://github.com/greenriver/hmis-warehouse/pull/4817/commits/055e3f4e4dec4a1754dccaa8792ec662443f6016

As an aside, while I was working on this, I couldn't find/remember a good test case for invalid client or enrollment. I tried some obvious ones (client DOB is 200 years ago, etc.) before realizing that [client validations are only run in specific contexts](https://github.com/greenriver/hmis-warehouse/blob/stable/drivers/hmis/app/models/hmis/hud/client.rb#L82), and external form submission is not one of them. I'm assuming that is intentional, because we want them to be able to create the client/enrollment and then fix DQ issues later.

Reply:
Oh, I didn't realize you were trying to catch errors from the FormProcessor run! method, too, but that makes sense. I guess I would suggest making the `rescue` block as small as possible to make that clearer, and to reduce the risk of swallowing errors that don't come from the form processor.

```
submissions.each do |record|
    next if already_processed

    begin
        run_form_processor!
    rescue StandardError => e
         # if there was an unexpected error running the form processor, it could mean the form response was malformed. catch this error so that it doesn't cause the whole bulk batch to fail.
          failed_to_review << e
    end

    if something_isnt_valid
       failed_to_review << record
       next
    end

end
```

And I agree it would be even better to subclass StandardError inside the form processor like you suggested, so we could `rescue FormProcessingError`. But I think that might involve a little too much code churn.. "unknown field error" isnt the only type of error, I think we would also wanna re-raise other exceptions as FormProcessingErrors [here](https://github.com/greenriver/hmis-warehouse/blob/5eed1631d1ae87951eb01bb1b4d4e7d064142b27/drivers/hmis/app/models/hmis/form/form_processor.rb#L80-L84).


In my example I didn't mean to say anything special about the validity check, sorry about that confusion. I agree that it makes sense to check the validity of the client/enrollment in order to prevent an unhandled exception on save!. One way to have and invalid client is to have a null PersonalID or EnrollmentID, but I'm not sure if there's even a way to make it happen with bad input.

Reply:
Ok, great, thank you for clarifying! I think this all makes sense.

> I'm not sure if there's even a way to make it happen with bad input.

I guess this makes me wonder whether we should just allow `.save!` to raise an unhandled exception? If it can't be due to bad data, it has to be because of a bug we introduced, so I think of it as in a different class of error -- and we want to get notified via sentry right away. On the other hand, I guess there's no harm in allowing the rest of the bulk submissions to succeed... as long as we never do [what I suggested in this comment](https://github.com/greenriver/hmis-warehouse/pull/4817#discussion_r1797320177) and turn off sentry alerts when it gets too noisy. I'll add some comments to this effect

Reply:
If theres no way to set it off from bad input, I agree it would be fine to let it raise. I didn't spend much time looking into it, though.